doc/design/prune.md

```diff
@@ -0,0 +1,110 @@
+# Prune
+
+## Motivation
+
+We want to support running inference, training and checkpointing in one `ProgramDesc`. We implement 
+`void Prune(const ProgramDesc* input, ProgramDesc* output)` function, which takes a `ProgramDesc`
+and generate a pruned `ProgramDesc`.
+
+## Challenge
+
+Pruning need to support both variables and operators being evaluation targets. Consider the following
+different situations.
+
+```python
+# Case 1: run foward pass.
+cost_np = session.run(target=cost)
+# Case 2: run backward passing.
+opts_np, _ = session.run(target=[cost, opt])
+# Case 3: run checkpointing
+_ = session.run(target=checkpoint)
+```
+
+## Solution
+
+To support evaluation of operators, we add `is_target` field in the `OpDesc`.
+
+```c++
+message OpDesc {
+  required string type = 3;
+  repeated Var inputs = 1;
+  repeated Var outputs = 2;
+  repeated Attr attrs = 4;
+  required bool is_target = 5 [ default = false ];
```

Comment:
When a client calls `sess.run(target=loss)`, PaddlePaddle will make a copy of `ProgramDesc` and set `is_target`

Reply:
So the `sess.run(target=loss)` is supposed to be in a `for` loop and it will take the same target repeated. We can optimize it by caching targets and their corresponding pruned ProgramDesc.

Reply:
right.